src/ncp/posix/infra_if.cpp

```diff
@@ -0,0 +1,168 @@
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define OTBR_LOG_TAG "INFRAIF"
+
+#include "infra_if.hpp"
+
+#include <ifaddrs.h>
+#ifdef __linux__
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#endif
+#include <netinet/icmp6.h>
+#include <sys/ioctl.h>
+
+#include "utils/socket_utils.hpp"
+
+namespace otbr {
+
+otbrError InfraIf::Dependencies::SetInfraIf(uint32_t                       aInfraIfIndex,
+                                            bool                           aIsRunning,
+                                            const std::vector<Ip6Address> &aIp6Addresses)
+{
+    OTBR_UNUSED_VARIABLE(aInfraIfIndex);
+    OTBR_UNUSED_VARIABLE(aIsRunning);
+    OTBR_UNUSED_VARIABLE(aIp6Addresses);
+
+    return OTBR_ERROR_NONE;
+}
+
+InfraIf::InfraIf(Dependencies &aDependencies)
+    : mDeps(aDependencies)
+    , mInfraIfIndex(0)
+{
+}
+
+void InfraIf::Init(void)
+{
+}
+
+void InfraIf::Deinit(void)
+{
+    mInfraIfIndex = 0;
+}
+
+otbrError InfraIf::SetInfraIf(const char *aIfName)
+{
+    otbrError               error = OTBR_ERROR_NONE;
+    std::vector<Ip6Address> addresses;
+
+    VerifyOrExit(aIfName != nullptr && strlen(aIfName) > 0, error = OTBR_ERROR_INVALID_ARGS);
+    VerifyOrExit(strnlen(aIfName, IFNAMSIZ) < IFNAMSIZ, error = OTBR_ERROR_INVALID_ARGS);
+    mInfraIfName = aIfName;
+
+    mInfraIfIndex = if_nametoindex(aIfName);
+    VerifyOrExit(mInfraIfIndex != 0, error = OTBR_ERROR_INVALID_STATE);
+
+    addresses = GetAddresses();
+
+    SuccessOrExit(mDeps.SetInfraIf(mInfraIfIndex, IsRunning(addresses), addresses), error = OTBR_ERROR_OPENTHREAD);
+exit:
+    otbrLogResult(error, "SetInfraIf");
+
+    return error;
+}
+
+bool InfraIf::IsRunning(const std::vector<Ip6Address> &aAddrs) const
+{
+    return mInfraIfIndex ? ((GetFlags() & IFF_RUNNING) && HasLinkLocalAddress(aAddrs)) : false;
+}
+
+uint32_t InfraIf::GetFlags(void) const
```

Comment:
why do we use uint32_t for this field? see the field [ifr_flags](https://man7.org/linux/man-pages/man7/netdevice.7.html#:~:text=sockaddr%20ifr_hwaddr%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20short-,ifr_flags,-%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20int%20%20%20%20%20%20%20%20%20%20%20%20%20ifr_ifindex%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20int)
```suggestion
short InfraIf::GetFlags(void) const
```